logsmanagement/flb_plugin.c

```diff
@@ -0,0 +1,1464 @@
+                    }
+
+                    /* NOTE/WARNING: 
+                     * kmsg timestamps are tricky. The timestamp will be 
+                     * *wrong** if the system has gone into hibernation since 
+                     * last boot and "p_file_info->use_log_timestamp" is set. 
+                     * Even if "p_file_info->use_log_timestamp" is NOT set, we 
+                     * need to use now_realtime_msec() as Fluent Bit timestamp
+                     * will also be wrong. */
+                    if( !strncmp(p->key.via.str.ptr, "sec", (size_t) p->key.via.str.size)){
+                        if(p_file_info->use_log_timestamp){
+                            timestamp += (now_realtime_sec() - now_boottime_sec() + p->val.via.i64) * MSEC_PER_SEC;
+                        }
+                        else if(!timestamp)
+                            timestamp = now_realtime_msec();
+                    }
+                    else if(!strncmp(p->key.via.str.ptr, "usec", (size_t) p->key.via.str.size) && 
+                            p_file_info->use_log_timestamp){
+                        timestamp += p->val.via.i64 / USEC_PER_MS;
+                    }
+                    else if(!strncmp(p->key.via.str.ptr, LOG_REC_KEY, (size_t) p->key.via.str.size)){
+                        message = (char *) p->val.via.str.ptr;
+                        message_size = p->val.via.str.size;
+
+                        m_assert(message, "message is NULL");
+                        m_assert(message_size, "message_size is 0");
+
+                        new_tmp_text_size += message_size + 1; // +1 for '\n'
+                    }
+                    else if(!strncmp(p->key.via.str.ptr, "priority", (size_t) p->key.via.str.size)){
+                        kmsg_sever = (int) p->val.via.u64;
+                    }
+                    ++p;
+                    continue;
+                } /* FLB_KMSG case end */
+
+                /* FLB_SYSTEMD or FLB_SYSLOG case */
+                if( p_file_info->log_type == FLB_SYSTEMD || 
+                    p_file_info->log_type == FLB_SYSLOG){
+                    if( p_file_info->use_log_timestamp && !strncmp( p->key.via.str.ptr, 
+                                                                    "_SOURCE_REALTIME_TIMESTAMP", 
+                                                                    (size_t) p->key.via.str.size)){
+                        m_assert(p->val.via.str.size <= TIMESTAMP_MS_STR_SIZE - 1, 
+                                "p->val.via.str.size >= TIMESTAMP_MS_STR_SIZE");
+                        strncpyz(timestamp_str, p->val.via.str.ptr, (size_t) p->val.via.str.size);
+
+                        /* TODO: Write dedicated function for timestamp conversion, with better error checking. */
+                        errno = 0;
+                        char *endptr = NULL;
+                        timestamp = strtoll(timestamp_str, &endptr, 10);
```

Comment:
Unless there is something really special here, please, prefer to call `str2ll`.

Reply:
Done, although it's not the same function as we don't check for out-of-range values. But there shouldn't be any here.